### PR TITLE
Check the time limit every (flush_interval / poll_timeout) number of rows from Kafka

### PR DESCRIPTION
**Pull request by abyss7**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description (up to few sentences):
Checking time-limit inside inferior streams allows to break the reading from Kafka consumer more frequently and to check the time limits for the top-level streams.

Detailed description (optional):
Reading from Kafka implies some polling timeout on read - it may drastically slow down the reading in a loop, if there is no time limit checks (like `BlockInputStreamFromRowInputStream::readImpl()`). Thus we give the option to tune this setting manually, and also we automatically tune the `rows_portion_size` to match the top-level time limit in the worst-case scenario when we read messages from Kafka on the edge of polling timeout.

### CODE REVIEW

**Comment by abyss7** [comment]
Should fix #2508

File: dbms/src/Formats/BlockInputStreamFromRowInputStream.h [hunk 1]
```diff
@@ -22,6 +22,7 @@ class BlockInputStreamFromRowInputStream : public IBlockInputStream
         const RowInputStreamPtr & row_input_,
         const Block & sample_,
         UInt64 max_block_size_,
+        UInt64 min_block_size_,
```
Comment (KochetovNicolai) [inline dbms/src/Formats/BlockInputStreamFromRowInputStream.h]: Name is not good. It's how often we check limits, and there is no connection with block size.

File: dbms/src/Formats/BlockInputStreamFromRowInputStream.h [hunk 2]
```diff
@@ -41,7 +42,7 @@ class BlockInputStreamFromRowInputStream : public IBlockInputStream
 private:
     RowInputStreamPtr row_input;
     Block sample;
-    UInt64 max_block_size;
+    UInt64 max_block_size, min_block_size;
```
Comment (KochetovNicolai) [inline dbms/src/Formats/BlockInputStreamFromRowInputStream.h]: Style. Declare on separate rows.

File: dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp [hunk 3]
```diff
@@ -39,7 +39,14 @@ void KafkaBlockInputStream::readPrefixImpl()
 
     buffer->subBufferAs<ReadBufferFromKafkaConsumer>()->subscribe(storage.topics);
 
-    addChild(FormatFactory::instance().getInput(storage.format_name, *buffer, storage.getSampleBlock(), context, max_block_size));
+    const auto & limits = getLimits();
+    const size_t poll_timeout = buffer->subBufferAs<ReadBufferFromKafkaConsumer>()->pollTimeout();
+    size_t min_block_size = std::min(max_block_size, limits.max_execution_time.totalMilliseconds() / poll_timeout);
```
Comment (KochetovNicolai) [inline dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp]: Possible division to 0 if `stream_poll_timeout_ms = 0`.

File: dbms/src/Storages/Kafka/StorageKafka.cpp [hunk 4]
```diff
@@ -210,8 +210,9 @@ BufferPtr StorageKafka::createBuffer()
     size_t batch_size = max_block_size;
     if (!batch_size)
         batch_size = settings.max_block_size.value;
+    size_t poll_timeout = settings.stream_poll_timeout_ms.totalMilliseconds();
 
-    return std::make_shared<DelimitedReadBuffer>(new ReadBufferFromKafkaConsumer(consumer, log, batch_size), row_delimiter);
+    return std::make_shared<DelimitedReadBuffer>(new ReadBufferFromKafkaConsumer(consumer, log, batch_size, poll_timeout), row_delimiter);
```
Comment (KochetovNicolai) [inline dbms/src/Storages/Kafka/StorageKafka.cpp]: Looks like possible memory leak. I think it's better to change `DelimitedReadBuffer` constructor (accept unique_ptr there).

**Comment by abyss7** [comment]
Also should partially fix #4308

**Pull request merged**